src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs

```diff
@@ -137,30 +135,34 @@ protected override void Dispose(bool disposing)
             }
 
             _context = null;
-            _handlers = null;
+            _handlers = Array.Empty<ExportLifetimeContext<IWorkspaceContextHandler>>();
         }
 
         protected override void Initialize()
         {
+            Assumes.NotNull(_context);
+
             _handlers = _workspaceContextHandlerFactories.Select(h => h.CreateExport())
                                                          .ToArray();
 
             foreach (ExportLifetimeContext<IWorkspaceContextHandler> handler in _handlers)
             {
-                handler.Value.Initialize(_context);
+                handler.Value.Initialize(_context!);
             }
         }
 
         private void ProcessProjectBuildFailure(IProjectRuleSnapshot snapshot)
         {
+            Assumes.NotNull(_context);
```

Comment:
Unfortunately these guard methods are not annotated to indicate that the argument is not null after this point. There is a `DoesNotReturnIf(bool)` attribute, but not a `DoesNotReturnIfNull(string parameterName)` attribute. It seems we have to `!` the first usage after this point. I filed https://github.com/dotnet/roslyn/issues/37071 to see if we could just `!` the argument to the guard method directly, which would be nicer.

Reply:
can we file an issue on the VSSDK folks for this?

Reply:
Can do, but there's nothing they can do until the compiler supports it. As soon as there's compiler support, I'll send them a PR if they don't get to it first.

Reply:
Actually it's possible to do this with the current attributes:

```c#
void AssertNotNull([NotNull] object? o) { }

string M(object? o)
{
    AssertNotNull(o);
    return o.ToString(); // no diagnostic here
}
```